module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,77 @@
+#include "advancedCalculator.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <cmath>
+#include <functional>
+#include <iterator>
+#include <map>
+
+#include <iostream>
+
+double factorial (double num) {
+    if (num <= 0) 
+        return 1;
+    return std::tgamma(num + 1);
+}
+
+const std::map<char, std::function<double(double, double)>> equation {
+    {'+', std::plus<double>()},
+    {'-', std::minus<double>()},
+    {'*', std::multiplies<double>()},
+    {'/', std::divides<double>()},
+    {'%', std::modulus<int>()},
+    {'^', [](double base, double exp) { return std::pow(base, exp); } },
+    {'$', [](double base, double exp) { return std::pow(base, 1.0/exp); } }
+};
```

Comment:
My suggestion:

You can put factorial there :~}. Let's try sth like that.

{"!", [](const double value, const double defaultValue) { return (value < 0) ? defaultValue : tgamma(value + defaultValue); }};

Value of variable defaultValue is "1". You can define this variable as constexpr for example.

Reply:
Was thinking about it, but that change more that just one line, and i still have to call factorial separatly couse of just one argument.

Reply:
I think you could insert here factorial. Just make separate function.
`{'!', [](double lhs, double rhs) { return factorial(lhs); }},`